### PR TITLE
[MXNET-379] L1 Normalization

### PR DESCRIPTION
**Pull request by anirudhacharya**
## Description ##
l1 norm operator

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- l1 norm

## Comments ##
- @haojin2 @eric-haibin-lin @piiswrong for review

### CODE REVIEW

File: src/operator/tensor/broadcast_reduce_op.h [hunk 1]
```diff
@@ -880,30 +880,34 @@ inline bool L2NormStorageType(const nnvm::NodeAttrs& attrs,
...
+    }
+    if (!dispatched && in_stype == kCSRStorage && axis.ndim() == 1 && !param.keepdims &&
+        (axis[0] == 0 || axis[0] == 1) && param.ord == 2) {
+      // l2 norm: csr, axis = 0/1 -> dns
+      dispatched = storage_type_assign(&out_stype, kDefaultStorage, dispatch_mode,
+                                       dispatch_ex);
+    }
+    if (!dispatched) {
+      dispatched = dispatch_fallback(out_attrs, dispatch_mode);
+    }
```
Comment (haojin2) [inline src/operator/tensor/broadcast_reduce_op.h]: You do not need the extra else branch for param.ord. Just do:
```c++
If (param.ord == 2) {
     // original storage type assign logic
}
if (!dispatched) {
      dispatched = dispatch_fallback(out_attrs, dispatch_mode);
}
```

File: tests/python/unittest/test_ndarray.py [hunk 2]
```diff
@@ -1304,6 +1304,36 @@ def test_norm(ctx=default_context()):
...
             mx.test_utils.assert_almost_equal(arr1, arr2.asnumpy())
 
+@with_seed()
+def test_l1norm(ctx=default_context()):
+    np_arr = np.random.uniform(size=(3, 3, 3, 3))
+    mx_arr = mx.nd.array(np_arr, ctx=ctx)
+    arr1 = np.sum(abs(np_arr), keepdims=False)
+    arr2 = mx.nd.norm(mx_arr, ord=1, keepdims=False)
+    print(arr1)
+    print(arr2.asnumpy())
```
Comment (haojin2) [inline tests/python/unittest/test_ndarray.py]: why print here?

File: tests/python/unittest/test_operator.py [hunk 3]
```diff
@@ -2879,6 +2879,36 @@ def npy_layer_norm(data, gamma, beta, axis=1, eps=1E-5):
                                grad_nodes={'data': req, 'gamma': req, 'beta': req},
                                numeric_eps=1e-2, rtol=1e-2, atol=1e-2)
 
+@with_seed()
+def test_norml1():
```
Comment (haojin2) [inline tests/python/unittest/test_operator.py]: better use "test_l1_normalization"

File: tests/python/unittest/test_operator.py [hunk 4]
```diff
@@ -2879,6 +2879,36 @@ def npy_layer_norm(data, gamma, beta, axis=1, eps=1E-5):
...
+            for keep_dims in [True, False]:
+                norm_sym = mx.symbol.norm(data=data, ord=1, axis=i, keepdims=keep_dims)
+                exe = norm_sym.simple_bind(ctx, data=(4,4,4,4))
+                exe.arg_dict['data'][:] = in_data
+                npy_out = np.sum(abs(in_data), axis=i, keepdims=keep_dims)
+                out = exe.forward()[0]
+                assert_almost_equal(out, npy_out, rtol=1e-2, atol=1e-5)
+                # check gradient
+                check_numeric_gradient(out, [in_data], numeric_eps=1e-3, rtol=1e-2, atol=1e-3)
+                if i < 3:
```
Comment (haojin2) [inline tests/python/unittest/test_operator.py]: your `i` is already in [0,4), why this extra check here?

File: tests/python/unittest/test_operator.py [hunk 5]
```diff
@@ -2879,6 +2879,36 @@ def npy_layer_norm(data, gamma, beta, axis=1, eps=1E-5):
...
+@with_seed()
+def test_norml1():
+    def npy_l1norm(data, axis):
+        np.sum(abs(np_arr), axis=i, keepdims=True)
+    ctx = default_context()
+    data = mx.symbol.Variable('data')
+
+    for dtype in [np.float16, np.float32, np.float64]:
+        dtype = np.float32
+        in_data = np.random.uniform(-1, 1, (4,4,4,4)).astype(dtype)
```
Comment (haojin2) [inline tests/python/unittest/test_operator.py]: Better randomize the shape of your test data here

File: tests/python/unittest/test_operator.py [hunk 6]
```diff
@@ -2879,6 +2879,36 @@ def npy_layer_norm(data, gamma, beta, axis=1, eps=1E-5):
...
 
+@with_seed()
+def test_norml1():
+    def npy_l1norm(data, axis):
+        np.sum(abs(np_arr), axis=i, keepdims=True)
+    ctx = default_context()
+    data = mx.symbol.Variable('data')
+
+    for dtype in [np.float16, np.float32, np.float64]:
+        dtype = np.float32
```
Comment (haojin2) [inline tests/python/unittest/test_operator.py]: With this line you are only testing float32 here

File: tests/python/unittest/test_operator.py (hunk 5 above)
Comment (haojin2) [reply tests/python/unittest/test_operator.py]: Also, better randomize the number of dimensions of your input.

File: tests/python/unittest/test_operator.py [hunk 7]
```diff
@@ -2879,6 +2879,36 @@ def npy_layer_norm(data, gamma, beta, axis=1, eps=1E-5):
...
+                assert_almost_equal(out, npy_out, rtol=1e-2, atol=1e-5)
+                # check gradient
+                check_numeric_gradient(out, [in_data], numeric_eps=1e-3, rtol=1e-2, atol=1e-3)
+                if i < 3:
+                    norm_sym = mx.symbol.norm(data=data, ord=1, axis=(i, i+1), keepdims=keep_dims)
+                    exe = norm_sym.simple_bind(ctx, data=(4,4,4,4))
+                    exe.arg_dict['data'][:] = in_data
+                    npy_out = np.sum(abs(in_data), axis=(i, i+1), keepdims=keep_dims)
+                    out = exe.forward()[0]
+                    assert_almost_equal(out, npy_out, rtol=1e-2, atol=1e-5)
```
Comment (haojin2) [inline tests/python/unittest/test_operator.py]: Everything you're doing above should be able to do in one check_symbolic_forward call

File: tests/python/unittest/test_operator.py [hunk 8]
```diff
@@ -2879,6 +2879,36 @@ def npy_layer_norm(data, gamma, beta, axis=1, eps=1E-5):
...
+        for i in range(4):
+            for keep_dims in [True, False]:
+                norm_sym = mx.symbol.norm(data=data, ord=1, axis=i, keepdims=keep_dims)
+                exe = norm_sym.simple_bind(ctx, data=(4,4,4,4))
+                exe.arg_dict['data'][:] = in_data
+                npy_out = np.sum(abs(in_data), axis=i, keepdims=keep_dims)
+                out = exe.forward()[0]
+                assert_almost_equal(out, npy_out, rtol=1e-2, atol=1e-5)
+                # check gradient
+                check_numeric_gradient(out, [in_data], numeric_eps=1e-3, rtol=1e-2, atol=1e-3)
```
Comment (haojin2) [inline tests/python/unittest/test_operator.py]: is rtol too lenient? Should you not have a tighter rtol/atol for fp32 and fp64?

File: src/operator/tensor/broadcast_reduce_op.h [hunk 9]
```diff
@@ -999,13 +1003,18 @@ void L2NormCompute(const nnvm::NodeAttrs& attrs,
...
-      ctx, inputs, req, outputs, small);
-  SqRootForL2<xpu>(ctx, req[0], outputs[0]);
+  if (param.ord == 1) {
+    ReduceAxesComputeImpl<xpu, mshadow::red::sum, false, mshadow_op::abs>(
+          ctx, inputs, req, outputs, small);
+  } else if (param.ord == 2) {
+    ReduceAxesComputeImpl<xpu, mshadow::red::sum, false, mshadow_op::square>(
+        ctx, inputs, req, outputs, small);
+    SqRootForL2<xpu>(ctx, req[0], outputs[0]);
+  }
```
Comment (haojin2) [inline src/operator/tensor/broadcast_reduce_op.h]: Better add an extra else branch with a warning message if the ord exceeds what we support.

File: src/operator/tensor/broadcast_reduce_op_value.cc [hunk 10]
```diff
@@ -322,7 +328,7 @@ NNVM_REGISTER_OP(_backward_norm)
   [](const NodeAttrs& attrs) {
     return std::vector<ResourceRequest>{ResourceRequest::kTempSpace};
   })
-.set_attr<FCompute>("FCompute<cpu>", L2NormGradCompute<cpu>);
+.set_attr<FCompute>("FCompute<cpu>", LpNormGradCompute<cpu>);
```
Comment (haojin2) [inline src/operator/tensor/broadcast_reduce_op_value.cc]: Should also update the gpu side

File: src/operator/tensor/broadcast_reduce_op.h (hunk 9 above)
Comment (anirudhacharya) [reply src/operator/tensor/broadcast_reduce_op.h]: there are constraint checks which will catch ord>2 before the code reaches this point.

File: tests/python/unittest/test_operator.py (hunk 6 above)
Comment (anirudhacharya) [reply tests/python/unittest/test_operator.py]: had put in for debugging something. will remove it and randomize the shape tuple with ``test_utils.rand_shape_nd``

File: tests/python/unittest/test_ndarray.py (hunk 2 above)
Comment (anirudhacharya) [reply tests/python/unittest/test_ndarray.py]: will remove it.

File: tests/python/unittest/test_operator.py (hunk 8 above)
Comment (anirudhacharya) [reply tests/python/unittest/test_operator.py]: what should the values be for fp64 and fp32?

Comment (haojin2) [reply tests/python/unittest/test_operator.py]: I think usually we have rtol=1e-4 and atol=1e-5

File: src/operator/tensor/broadcast_reduce_op.h [hunk 11]
```diff
@@ -880,30 +880,32 @@ inline bool L2NormStorageType(const nnvm::NodeAttrs& attrs,
...
+    if (!dispatched && in_stype == kCSRStorage && axis.ndim() == 1 && !param.keepdims &&
+        (axis[0] == 0 || axis[0] == 1) && param.ord == 2) {
+      // l2 norm: csr, axis = 0/1 -> dns
+      dispatched = storage_type_assign(&out_stype, kDefaultStorage, dispatch_mode,
+                                       dispatch_ex);
+    }
   }
   if (!dispatched) {
-    dispatched = dispatch_fallback(out_attrs, dispatch_mode);
+        dispatched = dispatch_fallback(out_attrs, dispatch_mode);
```
Comment (haojin2) [inline src/operator/tensor/broadcast_reduce_op.h]: Please fix the indentation.

File: src/operator/tensor/broadcast_reduce_op.h [hunk 12]
```diff
@@ -984,13 +986,13 @@ void SqRootForL2(const OpContext& ctx, OpReqType req, const TBlob &output) {
...
 template<typename xpu>
-void L2NormCompute(const nnvm::NodeAttrs& attrs,
+void LpNormCompute(const nnvm::NodeAttrs& attrs,
                    const OpContext& ctx,
                    const std::vector<TBlob>& inputs,
                    const std::vector<OpReqType>& req,
                    const std::vector<TBlob>& outputs) {
   const NormParam& param = nnvm::get<NormParam>(attrs.parsed);
-  CHECK_EQ(param.ord, 2) << "norm only support ord=2";
+  CHECK_LE(param.ord, 2) << "norm only supports ord=1 and ord=2";
```
Comment (haojin2) [inline src/operator/tensor/broadcast_reduce_op.h]: I think you should do "CHECK(param.ord == 1 || param.ord == 2)" as any ord <= 0 will also pass this check here.

File: tests/python/unittest/test_operator.py [hunk 13]
```diff
@@ -2879,6 +2879,32 @@ def npy_layer_norm(data, gamma, beta, axis=1, eps=1E-5):
...
+                # check gradient
+                #check_numeric_gradient(norm_sym, [in_data], numeric_eps=1e-3, rtol=1e-2, atol=1e-3)
+                if i < in_data_dim-1:
+                    norm_sym = mx.symbol.norm(data=data, ord=1, axis=(i, i+1), keepdims=keep_dims)
+                    npy_out = np.sum(abs(in_data), axis=(i, i+1), keepdims=keep_dims)
+                    check_symbolic_forward(norm_sym, [in_data], [npy_out],
+                                       rtol=1e-2 if dtype is np.float16 else 1e-5,
+                                       atol=1e-5, ctx=ctx)
+                    # check gradient
+                    #check_numeric_gradient(norm_sym, [in_data], numeric_eps=1e-3, rtol=1e-2, atol=1e-3)
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_operator.py]: why comment this out?

Comment (eric-haibin-lin) [reply tests/python/unittest/test_operator.py]: Also - check_symbolic_backward?

File: src/operator/tensor/broadcast_reduce_op.h [hunk 14]
```diff
@@ -880,27 +880,29 @@ inline bool L2NormStorageType(const nnvm::NodeAttrs& attrs,
...
-  if (!dispatched && in_stype == kCSRStorage && axis.ndim() == 1 && !param.keepdims &&
-      (axis[0] == 0 || axis[0] == 1) && param.ord == 2) {
-    // l2 norm: csr, axis = 0/1 -> dns
-    dispatched = storage_type_assign(&out_stype, kDefaultStorage, dispatch_mode,
-                                     dispatch_ex);
+    if (!dispatched && in_stype == kDefaultStorage) {
+      // dns -> dns
+      dispatched = storage_type_assign(&out_stype, kDefaultStorage, dispatch_mode,
+                                       DispatchMode::kFCompute);
+    }
```
Comment (haojin2) [inline src/operator/tensor/broadcast_reduce_op.h]: Please take the default case out of the if branch:
```c++
if (!dispatched && in_stype == kDefaultStorage) {
  // ...
}
if (param.ord == 2) {
   // the rest cases
}
if (!dispatched) {
  // fallback
}
```

**Comment by anirudhacharya** [comment]
@haojin2 @eric-haibin-lin @piiswrong ping for review. 

The ci failure is due to unrelated flaky test. Will retrigger the ci.

File: src/operator/tensor/broadcast_reduce_op.h [hunk 15]
```diff
@@ -1021,8 +1023,38 @@ void L2NormGradCompute(const nnvm::NodeAttrs& attrs,
   } else {
     small = ReduceAxesShapeImpl(outputs[0].shape_, param.axis, true, false);
   }
-  ReduceAxesBackwardUseInOutImpl<xpu, mshadow_op::div, false>(ctx, small, inputs,
+  if (param.ord == 1) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
```
Comment (haojin2) [inline src/operator/tensor/broadcast_reduce_op.h]: These to `using namespace` line are redundant.

File: src/operator/tensor/broadcast_reduce_op.h [hunk 16]
```diff
@@ -1021,8 +1023,38 @@ void L2NormGradCompute(const nnvm::NodeAttrs& attrs,
...
+          inputs[0].get_with_shape<xpu, ndim, DType>(dst_shape.get<ndim>(), s);
+        Tensor<xpu, ndim, DType> data =
+          inputs[1].get_with_shape<xpu, ndim, DType>(src_shape.get<ndim>(), s);
+        ASSIGN_DISPATCH(igrad, req[0],
+          broadcast_to(ograd, src_shape)*F<mshadow_op::sign>(data));
+      }
+    });
+  } else if (param.ord == 2) {
+    ReduceAxesBackwardUseInOutImpl<xpu, mshadow_op::div, false>(ctx, small, inputs,
                                                               req, outputs);
```
Comment (haojin2) [inline src/operator/tensor/broadcast_reduce_op.h]: Fix the alignment here.

File: src/operator/tensor/broadcast_reduce_op_value.cc [hunk 17]
```diff
@@ -284,10 +284,16 @@ array.
...
+        [5,6]]]
 
-  norm(x) = [5.47722578]
+  norm(x, ord=2, axis=1) = [[3.1622777 4.472136 ]
+                            [5.3851647 6.3245554]]
+
+  norm(x, ord=1, axis=1) = [[4., 6.],
+                            [7., 8.]]
 
   rsp = x.cast_storage('row_sparse')
```
Comment (haojin2) [inline src/operator/tensor/broadcast_reduce_op_value.cc]: Please also add specification for sparse-related cases below.

File: src/operator/tensor/broadcast_reduce_op_value.cc [hunk 18]
```diff
@@ -284,10 +284,16 @@ array.
 
 Examples::
 
-  x = [[1, 2],
-       [3, 4]]
+  x = [[[1,2],
+        [3,4]],
+       [[2,2],
+        [5,6]]]
```
Comment (haojin2) [inline src/operator/tensor/broadcast_reduce_op_value.cc]: Please insert spaces just like the format before.

File: src/operator/tensor/broadcast_reduce_op.h (hunk 16 above)
Comment (anirudhacharya) [reply src/operator/tensor/broadcast_reduce_op.h]: it is aligned properly, it even passed the cpplint.

File: tests/python/unittest/test_operator.py [hunk 19]
```diff
@@ -3009,6 +3009,49 @@ def npy_layer_norm(data, gamma, beta, axis=1, eps=1E-5):
...
+        return np.linalg.norm(input_data, axis=axis, keepdims=keepdims)
+
+    ctx = default_context()
+    data = mx.symbol.Variable('data')
+    in_data_dim = random_sample([4,5,6], 1)[0]
+    in_shape = rand_shape_nd(in_data_dim)
+    for order in [1, 2]:
+        for dtype in [np.float16, np.float32, np.float64]:
+            in_data = np.random.uniform(-1, 1, in_shape).astype(dtype)
+            in_data[abs(in_data) < 1e-2] = 1e-2
```
Comment (haojin2) [inline tests/python/unittest/test_operator.py]: Please only do this for abs(in_data) < eps, where eps is the eps you're using for check_numeric_gradient

File: src/operator/tensor/broadcast_reduce_op.h [hunk 20]
```diff
@@ -880,27 +880,24 @@ inline bool L2NormStorageType(const nnvm::NodeAttrs& attrs,
...
+    }
+    if (!dispatched && in_stype == kCSRStorage && axis.ndim() == 1 && !param.keepdims &&
+        (axis[0] == 0 || axis[0] == 1) && param.ord == 2) {
+      // l2 norm: csr, axis = 0/1 -> dns
+      dispatched = storage_type_assign(&out_stype, kDefaultStorage, dispatch_mode,
+                                       dispatch_ex);
+    }
   }
   if (!dispatched) {
     dispatched = dispatch_fallback(out_attrs, dispatch_mode);
```
Comment (eric-haibin-lin) [inline src/operator/tensor/broadcast_reduce_op.h]: Do not dispatch to fallback if all your inputs are dense.

File: src/operator/tensor/broadcast_reduce_op_value.cc (hunk 17 above)
Comment (anirudhacharya) [reply src/operator/tensor/broadcast_reduce_op_value.cc]: I do not understand what you mean by "specification for sparse-related cases" could you please elaborate?

File: tests/python/unittest/test_operator.py [hunk 21]
```diff
@@ -3009,6 +3009,49 @@ def npy_layer_norm(data, gamma, beta, axis=1, eps=1E-5):
...
+            for i in range(in_data_dim):
+                norm_sym = mx.symbol.norm(data=data, ord=order, axis=i, keepdims=True)
+                npy_out = l1norm(in_data, i) if order==1 else l2norm(in_data, i)
```
Comment (haojin2) [inline tests/python/unittest/test_operator.py]: maybe use `is` instead of `==`

File: src/operator/tensor/broadcast_reduce_op_value.cc (hunk 10 above)
Comment (anirudhacharya) [reply src/operator/tensor/broadcast_reduce_op_value.cc]: this is done

File: src/operator/tensor/broadcast_reduce_op_value.cc (hunk 18 above)
Comment (anirudhacharya) [reply src/operator/tensor/broadcast_reduce_op_value.cc]: i will add the spaces, but I borrowed the format here from ``sum`` operator in line ~92 and they dont have spaces.

File: src/operator/tensor/broadcast_reduce_op.h (hunk 15 above)
Comment (anirudhacharya) [reply src/operator/tensor/broadcast_reduce_op.h]: shall remove it.

File: src/operator/tensor/broadcast_reduce_op.h (hunk 20 above)
Comment (haojin2) [reply src/operator/tensor/broadcast_reduce_op.h]: @anirudhacharya I think you missed the default case when copying the previous code over.

File: src/operator/tensor/broadcast_reduce_op.h (hunk 16 above)
Comment (haojin2) [reply src/operator/tensor/broadcast_reduce_op.h]: No, you should have `req` aligned with `ctx` above.

File: src/operator/tensor/broadcast_reduce_op_value.cc (hunk 17 above)
Comment (haojin2) [reply src/operator/tensor/broadcast_reduce_op_value.cc]: the lines below have examples for sparse, and you should update them with the latest changes to the behavior of this op, you should also add that only "ord=2" supports sparse ndarrays after your changes.

File: tests/python/unittest/test_operator.py (hunk 21 above)
Comment (anirudhacharya) [reply tests/python/unittest/test_operator.py]: python's ``is`` compares object reference whereas == compares for value. Here we want to compare the values and not object references. Using ``is`` might make the test fail. 

For example - 
```python
>>> 625 is 25**2
False
>>> 625 == 25**2
True
```

File: src/operator/tensor/broadcast_reduce_op_value.cc (hunk 18 above)
Comment (haojin2) [reply src/operator/tensor/broadcast_reduce_op_value.cc]: Then would you please fix those too?

File: tests/python/unittest/test_operator.py (hunk 19 above)
Comment (anirudhacharya) [reply tests/python/unittest/test_operator.py]: sure, i will use 1e-3. that is the epsilon i am using in check_gradient

File: src/operator/tensor/broadcast_reduce_op_value.cc (hunk 18 above)
Comment (anirudhacharya) [reply src/operator/tensor/broadcast_reduce_op_value.cc]: sure i can do that

File: tests/python/unittest/test_operator.py (hunk 21 above)
Comment (haojin2) [reply tests/python/unittest/test_operator.py]:
```python
>>> a = 1
>>> a is 1
True
```

Comment (anirudhacharya) [reply tests/python/unittest/test_operator.py]: https://docs.python.org/2/library/operator.html#operator.is_ ``is`` compares objects and not values.

File: src/operator/tensor/broadcast_reduce_op.h (hunk 20 above)
Comment (anirudhacharya) [reply src/operator/tensor/broadcast_reduce_op.h]: i will fix this

File: tests/python/unittest/test_operator.py (hunk 21 above)
Comment (haojin2) [reply tests/python/unittest/test_operator.py]: Okay, so maybe you can go ahead and change all
```Python
1e-x if dtype is np.float16 else 1e-y
```
in this function to use `==`?

File: src/operator/tensor/broadcast_reduce_op_value.cc (hunk 17 above)
Comment (anirudhacharya) [reply src/operator/tensor/broadcast_reduce_op_value.cc]: fixed

File: tests/python/unittest/test_operator.py [hunk 22]
```diff
@@ -3009,6 +3009,50 @@ def npy_layer_norm(data, gamma, beta, axis=1, eps=1E-5):
...
+                npy_out_backward = np.sign(in_data) if order==1 else in_data/npy_out 
+                check_symbolic_forward(norm_sym, [in_data], [npy_out],
+                                        rtol=1e-2 if dtype is np.float16 else 1e-5,
+                                        atol=1e-2 if dtype is np.float16 else 1e-5, ctx=ctx)
+                check_symbolic_backward(norm_sym, [in_data], [np.ones(npy_out.shape)],
+                                        [npy_out_backward],
+                                        rtol=1e-2 if dtype is np.float16 else 1e-5,
+                                        atol=1e-2 if dtype is np.float16 else 1e-5, ctx=ctx)
+                # check gradient
+                check_numeric_gradient(norm_sym, [in_data], numeric_eps=epsilon, rtol=1e-2, atol=1e-3)
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_operator.py]: rtol and atol at line 3040 and 3053 seems high. Is this intended?

Comment (anirudhacharya) [reply tests/python/unittest/test_operator.py]: yes. check line 2963. do you suggest other values?

**Comment by anirudhacharya** [comment]
@eric-haibin-lin can you please merge this.

**Pull request merged**